### PR TITLE
gonum/plot: add support for inverting an axis

### PR DESCRIPTION
**Pull request by npotts**
This change modifies gonum/plot to add direct library support to invert an axis

This change was spurned by attempting to plot atmospheric data, which is
commonly plotted against an inverted pressure axis such as this [example](https://www.ncl.ucar.edu/Applications/Images/h_lat_1_lg.png).   At the suggestion of Sebastion Binet I am submitting this patch. This is somewhat related to #316.

Signed-off-by: npotts <npotts@users.noreply.github.com>

Please take a look.

### CODE REVIEW

File: plotter/invertedaxis_example_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	p.Add(f, plotter.NewGrid())
+	p.Legend.Add("exp(x)", f)
+
+	//Notice that both .Min and .Max for X and Y are both in 'normal' order
+	p.X.Min = f.XMin
+	p.X.Max = f.XMax
+	p.Y.Min = math.Exp(f.XMin)
+	p.Y.Max = math.Exp(f.XMax)
+
+	err = p.Save(10*vg.Centimeter, 10*vg.Centimeter, "testdata/invertedlogscale.png")
```
Comment (sbinet) [inline plotter/invertedaxis_example_test.go]: could you add a test (so we make sure everything is alright) ?
Examples without a:
```go
// Output:
// blabla
```
aren't run as part of `go test`...

something like that perhaps:
```go
func TestInvertedScale(t *testing.T) {
	cmpimg.CheckPlot(Example_invertedScale, t, "invertedscale.png")
}
```

you'll probably have to add the png file :)

File: plotter/invertedaxis_example_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+// Example_invertedScale shows how to create a plot with an inverted Y-axis.
+// This is nearly identical to the Log example, except it inverts the Y axis.
+func Example_invertedScale() {
+	p, err := plot.New()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	p.Title.Text = "Exampled of an Inverted Axis"
```
Comment (sbinet) [inline plotter/invertedaxis_example_test.go]: `s/Exampled/Example/`
`s/an Inverted Axis/inverted axes/` (as both x and y are inverted)

File: plotter/invertedaxis_example_test.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	f := plotter.NewFunction(math.Exp)
+	f.XMin = 0.2
+	f.XMax = 10
+
+	f.Color = color.RGBA{R: 255, A: 255}
+
+	p.Add(f, plotter.NewGrid())
+	p.Legend.Add("exp(x)", f)
+
+	//Notice that both .Min and .Max for X and Y are both in 'normal' order
```
Comment (sbinet) [inline plotter/invertedaxis_example_test.go]: `s|//Notice|// Notice|`

File: plotter/invertedaxis_example_test.go (hunk 1 above)
Comment (npotts) [reply plotter/invertedaxis_example_test.go]: Ok I think I did it right.  Let me know otherwise

File: axis.go [hunk 4]
```diff
@@ -171,6 +171,18 @@ func (LogScale) Normalize(min, max, x float64) float64 {
 	return (log(x) - logMin) / (log(max) - logMin)
 }
 
+// InvertedScale can be used as the value of an Axis.Scale function to
+// invert the axis using any Normalizer. This can be useful for plots that
```
Comment (ctessum) [inline axis.go]: The second sentence here should be deleted, I think. It's not adding information

File: plotter/invertedaxis_example_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	"testing"
+
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/internal/cmpimg"
+	"gonum.org/v1/plot/plotter"
+	"gonum.org/v1/plot/vg"
+)
+
+// Example_invertedScale shows how to create a plot with an inverted Y-axis.
+// This is nearly identical to the Log example, except it inverts the Y axis.
```
Comment (ctessum) [inline plotter/invertedaxis_example_test.go]: I think the second sentence should be deleted, and the first sentence changed to say that both axes are inverted.

**Comment by npotts** [comment]
Ill change those ASAP.

Are you all open to the addition of some limited number of "helpers" like this?  I have some basic helper functions that aid in labeling, enforcing a minimal number of ticks and some other useful helper structs that I have been abusing quite heavily.

File: plotter/invertedaxis_example_test.go (hunk 5 above)
Comment (npotts) [reply plotter/invertedaxis_example_test.go]: 👍

File: axis.go (hunk 4 above)
Comment (npotts) [reply axis.go]: 👍

File: axis.go [hunk 6]
```diff
@@ -171,6 +171,17 @@ func (LogScale) Normalize(min, max, x float64) float64 {
 	return (log(x) - logMin) / (log(max) - logMin)
 }
 
+// InvertedScale can be used as the value of an Axis.Scale function to
+// invert the axis using any Normalizer.
+type InvertedScale struct{ Normalizer }
+
+var _ Normalizer = InvertedScale{}
+
+// Normalize returns a normalized [0, 1] value for the position of x
```
Comment (kortschak) [inline axis.go]: Final dot.

File: plotter/invertedaxis_example_test.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
+// Copyright ©2017 The Gonum Authors. All rights reserved.
```
Comment (kortschak) [inline plotter/invertedaxis_example_test.go]: 2018

File: plotter/invertedaxis_example_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	f := plotter.NewFunction(math.Exp)
+	f.XMin = 0.2
+	f.XMax = 10
+
+	f.Color = color.RGBA{R: 255, A: 255}
+
+	p.Add(f, plotter.NewGrid())
+	p.Legend.Add("exp(x)", f)
+
+	// Notice that both .Min and .Max for X and Y are both in 'normal' order
```
Comment (kortschak) [inline plotter/invertedaxis_example_test.go]: s/Notice that b/B/ and final dot. I think something other than 'normal' would be better here.

File: plotter/invertedaxis_example_test.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	"math"
+	"testing"
+
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/internal/cmpimg"
+	"gonum.org/v1/plot/plotter"
+	"gonum.org/v1/plot/vg"
+)
+
+// Example_invertedScale shows how to create a plot with an inverted axes.
```
Comment (kortschak) [inline plotter/invertedaxis_example_test.go]: This comment will not be rendered. If you want commentary in an example, it must be in the body. In that case you don't need to use godoc phrasing, just normal prose is OK.

Comment (npotts) [reply plotter/invertedaxis_example_test.go]: I don't write too many Test Examples - Its just means I need more practice

I moved the comment into the body as my intent was to provide more prose on how to exactly use it. gonum/plot has a particular cadence/style/feel when compared to other plotting libraries - I feel that more hints as to design patterns should make first time users get moving a bit faster.

File: axis.go (hunk 6 above)
Comment (npotts) [reply axis.go]: ✔️

File: plotter/invertedaxis_example_test.go (hunk 9 above)
Comment (kortschak) [reply plotter/invertedaxis_example_test.go]: Yeah, no worries. You're doing good work and sometime the vagaries of godoc rendering a difficult (I usually have a godoc server running and check what the docs look like to make sure everything is good, particularly when examples are part of the docs.).

**Comment by sbinet** [comment]
needs https://github.com/gonum/gonum/pull/726

File: plotter/invertedaxis_example_test.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	"testing"
+
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/internal/cmpimg"
+	"gonum.org/v1/plot/plotter"
+	"gonum.org/v1/plot/vg"
+)
+
+func Example_invertedScale() {
+	// This samples is nearly identical to the LogScale, other than
```
Comment (kortschak) [inline plotter/invertedaxis_example_test.go]: "This example"?

**Comment by sbinet** [comment]
thanks again and welcome

**Comment by kortschak** [comment]
This needed fixes in the commit message.

**Pull request merged**